PMM-4846 Restore CLI flags.

percona-csalguero opened the pull request:
Added global flags to add_xxx inventory/services management add commands:
- --host
- --port
- --service-name

Order of priority for Flags and Arguments will be:
- Flags
- Arguments

Example:
```
pmm-admin add mysql --port=1234 mymysql 127.0.0.1:3306
```
will look for MySql at `127.0.0.1:1234`

percona-csalguero committed:
PMM-4846 WIP - Switching branches

percona-csalguero committed:
PMM-4846 Restore cli options

Added global flags to add_xxx inventory/services management add
commands:
- --host
- --port
- --service-name

Order of priority for Flags and Arguments will be:
- Flags
- Arguments

Example:
```
pmm-admin add mysql --port=1234 mymysql 127.0.0.1:3306
```
will look for MySql at `127.0.0.1:1234`

BupycHuk commented on commands/inventory/inventory.go:
```diff
@@ -29,4 +29,38 @@ var (
+	// Add command gobal flags
+	AddAddressFlag     = inventoryAddC.Flag("host", "Hostname").String()
+	AddPortFlag        = inventoryAddC.Flag("port", "Port number").Int64()
+	AddServiceNameFlag = inventoryAddC.Flag("service-name", "Service name").String()
```
it'll affect agents and nodes commands. let's do it for `addServiceC` and move to `add_service.go`

BupycHuk commented on commands/inventory/inventory.go:
```diff
@@ -29,4 +29,38 @@ var (
+// - addServiceProxySQLCommand
+// - addServiceMongoDBCommand
+// - ddServicePostgreSQLCommand
+func processGlobalAddFlags(cmd getter) (string, string, int64) {
```
@AlekSi, @rnovikovP  do we need this feature for inventory commands?

AlekSi replied:
It is not _required_ for this ticket. But I would prefer to move away from positional arguments in most places – practice showed that they are too confusing.
So that depends on how complicated implementation is.

AlekSi replied:
In the end, I will revert changes in inventory commands – this PR implements them only for Services, leaving Nodes and Agents with only positional arguments. We need to think more about them.

percona-csalguero replied:
I made it this way because of this comment:
https://github.com/percona/pmm-admin/pull/86#discussion_r358405095

AlekSi replied:
Both are correct. We don't want ports for Nodes, but we also don't want replacements for positional arguments only for some commands, and not for others.

BupycHuk commented on commands/management/add.go:
```diff
@@ -16,10 +16,54 @@
+	}
+
+	addressPort := cmd.GetAddressPort()
+	if *AddAddressFlag != "" && *AddPortFlag != 0 {
```
AFAIK, each of them should be able to override separately. @AlekSi @rnovikovP?

AlekSi replied:
Correct. `--port=12345 1.2.3.4` should set full address to `1.2.3.4:12345`, not `:12345`.

AlekSi commented on commands/management/add.go:
```diff
@@ -16,10 +16,54 @@
+
+	addressPort := cmd.GetAddressPort()
+	if *AddAddressFlag != "" && *AddPortFlag != 0 {
+		addressPort = fmt.Sprintf("%s:%d", *AddAddressFlag, *AddPortFlag)
```
That will not work correctly for IPv6. Use `net.JoinHostPort`.

percona-csalguero committed:
PMM-4846 Restore cli options

- Address, port and service name flags has been moved from Add command
to Add service command so only adding services will have this new flags.
- Added better tests to inventory and management pkgs

AlekSi committed:
Merge branch 'master' into PMM-4846_restore_cli_options

AlekSi committed:
PMM-4846 Un-export variables.

AlekSi committed:
PMM-4846 Fix typo.

AlekSi committed:
PMM-4846 Tweak help output.

AlekSi committed:
PMM-4846 Use consistent names.

AlekSi committed:
PMM-4846 Remove inventory commands changes.

AlekSi committed:
PMM-4846 Minor cleanups.
